server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                                .collect(Collectors.toList())
+                        )
+                        .collect(Collectors.toList())
+                );
+            }
+            List<NodeAllocationResult> nodeAllocationResults = allocateDecision.getNodeDecisions();
+            boolean diagnosisFound = checkIsAllocationDisabled(diagnosisOutput, shardRouting, nodeAllocationResults);
+            IndexMetadata index = state.metadata().index(shardRouting.index());
+            if (index != null) {
+                diagnosisFound |= checkDataTierRelatedIssues(diagnosisOutput, index, shardRouting, nodeAllocationResults);
+            }
+            if (diagnosisFound == false) {
+                diagnosisOutput.accept(ACTION_CHECK_ALLOCATION_EXPLAIN_API, shardRouting);
+            }
+        }
+    }
+
+    private static Predicate<NodeAllocationResult> nodeHasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        if (nodeAllocationResults.stream().allMatch(nodeHasDeciderResult(EnableAllocationDecider.NAME, Decision.Type.NO))) {
+            diagnosisOutput.accept(ACTION_ENABLE_ALLOCATIONS, shardRouting);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean checkDataTierRelatedIssues(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        IndexMetadata indexMetadata,
+        ShardRouting shardRouting,
+        List<NodeAllocationResult> nodeAllocationResults
+    ) {
+        boolean diagnosisFound = false;
+        if (indexMetadata.getTierPreference().size() > 0) {
+            List<NodeAllocationResult> dataTierNodes = nodeAllocationResults.stream()
+                .filter(nodeHasDeciderResult(DATA_TIER_ALLOCATION_DECIDER_NAME, Decision.Type.YES))
+                .collect(Collectors.toList());
+            if (dataTierNodes.isEmpty()) {
+                diagnosisOutput.accept(ACTION_ENABLE_TIERS, shardRouting);
+                diagnosisFound = true;
+            } else {
+                // All tier nodes at shards limit?
+                if (dataTierNodes.stream().allMatch(nodeHasDeciderResult(ShardsLimitAllocationDecider.NAME, Decision.Type.NO))) {
+                    diagnosisOutput.accept(ACTION_SHARD_LIMIT, shardRouting);
+                    diagnosisFound = true;
+                }
+
+                // All tier nodes conflict with allocation filters?
+                if (dataTierNodes.stream().allMatch(nodeHasDeciderResult(FilterAllocationDecider.NAME, Decision.Type.NO))) {
```

Comment:
I believe this is too broad? Would this catch the `require._id` filtering we use when shrinking for example?
Also I think we can have the reverse case too, where the `require.data` filtering says YES and but the target tier says `NO` (?) 

I have a feeling we'd be better with simulating a migration and detecting if any index in the cluster needs to be migrated ie. `MigrateToDataTiersRouting#migrateToDataTiersRouting( ..., dryRun true)`. What do you think? 

@dakrone

Reply:
I agree, giving it a dry run through the migrate action would be good, but it does live in the ILM project. It would either need to be moved to server, have its logic duplicated, or a simpler check performed.

Reply:
> I believe this is too broad? Would this catch the `require._id` filtering we use when shrinking for example?

I think this is too broad also, I wonder if instead of making this a blanket "filter allocation said NO" thing we could treat it similar to the migrate API where we special-case the "data" attribute? I don't think we want to say "run the migrate API" when an index happens to be yellow due to `_id` pinning required for the shrink API.

Reply:
I don't know what granularity we have in the code, but we might want to add another check for "oops you fat-fingered a filtered attribute and there is no node in the cluster that has that attribute" if possible. (not necessarily in this PR, maybe in future work?)

Reply:
Added [c671eb5](https://github.com/elastic/elasticsearch/pull/85071/commits/c671eb5b0da7c3ae963637e8b509a68c4ab30875) to not only check for old data attribute filters but also to see if they keep nodes from being selected (since their existence does not automatically explain an unassigned shard)

Reply:
++ to the "no nodes match filter" check, but definitely should be after this lands.